Core/COEditingContext.m

```diff
@@ -310,6 +325,105 @@ - (void)undeletePersistentRoot: (COPersistentRoot *)aPersistentRoot
     {
         [_persistentRootsPendingUndeletion addObject: aPersistentRoot];
     }
+
+	[self updateCrossPersistentRootReferencesToPersistentRoot: aPersistentRoot
+	                                                   branch: nil
+	                                                isDeleted: NO];
+	
+	/* Once all the dead references are fixed, remove them from the cache */
+	
+	NSSet * allBranches = [aPersistentRoot.branches setByAddingObjectsFromSet: aPersistentRoot.deletedBranches];
+	
+	for (COBranch *branch in allBranches)
+	{
+		[_deadRelationshipCache removePath: [COPath pathWithPersistentRoot: aPersistentRoot.UUID
+		                                                            branch: branch.UUID]];
+	}
+	[_deadRelationshipCache removePath: [COPath pathWithPersistentRoot: aPersistentRoot.UUID]];
+}
+
+/**
+ * When isDeleted is YES, turns live references to the given persistent root 
+ * into dead ones in referring persistent roots.
+ *
+ * When isDeleted is NO, turns dead references to the given persistent root
+ * into live ones in referring persistent roots.
+ *
+ * When isDeleted is YES and branch is nil, this means the persistent root 
+ * itself is deleted. In this case, we consider the branches to be all 
+ * transively deleted for this method logic, although there aren't marked as 
+ * deleted.
+ *
+ * The implementation must take in account deletion/undeletion can be:
+ *
+ * <list>
+ * <item>explicit with COPersistentRoot and COBranch API</item>
+ * <item>implicit deletion/undeletion when reloading persistent roots or branches 
+ * (e.g. isTargetDeletion comment).</item>
+ * </list>
+ *
+ * IMPORTANT: This method must never result in new changes to be committed, 
+ * since it replaces object references in memory, but doesn't change any item
+ * graphs (the references inside each COItem remain identical). Requiring no 
+ * commits means the references can be fixed in both tracking and current branches.
+ *
+ * See also -[COPersistentRoot setCurrentBranchObjectGraphToRevisionUUID:persistentRootUUID:].
+ */
+- (void)updateCrossPersistentRootReferencesToPersistentRoot: (COPersistentRoot *)aPersistentRoot
+                                                     branch: (COBranch *)aBranch
+                                                  isDeleted: (BOOL)isDeletion
+{
+	NSParameterAssert(aPersistentRoot != nil);
+	// NOTE: -delete/undeleteBranch: enforce !aBranch.isCurrentBranch already.
+	NSParameterAssert(aBranch == nil
+		|| (!aBranch.isCurrentBranch && aBranch.persistentRoot == aPersistentRoot));
+
+	for (COPersistentRoot *persistentRoot in [_loadedPersistentRoots objectEnumerator])
+	{
+		if (persistentRoot == aPersistentRoot)
+			continue;
+		
+		// TODO: Use -objectGraphWithoutUnfaulting or -allObjectGraphContexts to
+		// prevent loading every object graph contexts we check
```

Comment:
Initially I wrote it to unfault all object graphs to make tests easier to write and truly exhaustive. If we don't unfault all object graphs requiring reference fixes in -updateCrossPersistentRootReferencesToPersistentRoot:branch:isDeleted:, we must do it in the tests themselves to ensure they remain exhaustive. That's what we should truly do.

To reply to your last question, the code checks every branch (in every persistent root) to fix references. In -[COObjectGraphContext replaceObject:withObject:], I check dead reference cache/relationship cache and return immediately when the object graph is not concerned, otherwise I call -[COObject replaceReferencesToObjectIdenticalTo:withObject:]. 
As you suggest, it is surely possible to move this check to -updateCrossPersistentRootReferencesToPersistentRoot:branch:isDeleted:. I haven't really thought about it, since the most expansive part is to iterate over all the inner objects and we skip this one already. The other thing is that I would wait until lazy loading of persistent roots is implemented, since this could require other changes to the current implementation. Once we support lazy loading, iterating over all branches should be less costly too.
This puts aside, moving the check to -updateCrossPersistentRootReferencesToPersistentRoot:branch:isDeleted: would make the code simpler and easier to debug too.